### PR TITLE
Exit early on error

### LINKED ISSUE #4153: Failure to initialize merge plugin should be a critical error
**Describe the bug**

Same story as
- https://github.com/NethermindEth/nethermind/issues/4149
- https://github.com/NethermindEth/nethermind/issues/4150

Nethermind logged

```
el_1  | 2022-06-12 13:09:42.3302|Failed to initialize plugin Merge by Nethermind System.FormatException: One of the identified items was in an invalid format.
el_1  |    at Nethermind.Int256.UInt256.Parse(String value) in /src/int256/src/Nethermind.Int256/UInt256.cs:line 2027
el_1  |    at Nethermind.Merge.Plugin.IMergeConfig.get_TerminalTotalDifficultyParsed() in /src/Nethermind/Nethermind.Merge.Plugin/IMergeConfig.cs:line 45
el_1  |    at Nethermind.Merge.Plugin.PoSSwitcher.Initialize() in /src/Nethermind/Nethermind.Merge.Plugin/PoSSwitcher.cs:line 90
el_1  |    at Nethermind.Merge.Plugin.PoSSwitcher..ctor(IMergeConfig mergeConfig, ISyncConfig syncConfig, IDb metadataDb, IBlockTree blockTree, ISpecProvider specProvider, ILogManager logManager) in /src/Nethermind/Nethermind.Merge.Plugin/PoSSwitcher.cs:line 83
el_1  |    at Nethermind.Merge.Plugin.MergePlugin.Init(INethermindApi nethermindApi) in /src/Nethermind/Nethermind.Merge.Plugin/MergePlugin.cs:line 84
el_1  |    at Nethermind.Init.Steps.InitializePlugins.Execute(CancellationToken cancellationToken) in /src/Nethermind/Nethermind.Init/Steps/InitializePlugins.cs:line 48
```

and continue happily, but the merge plugin was not loaded. So when the CL found the merge block everything errored because:

```
cl_1  | Eph 1/2 5.869[ETH1]             info: Terminal POW block found! hash=0x4ebd4a511161d0ec70c6674395c8b9f06cfce7e120b12379c1ec2640f2fe7384, number=11, totalDifficulty=3743106036130323098097120681749119449189
el_1  | 2022-06-12 13:10:49.9308|Waiting for peers... 7s 
el_1  | 2022-06-12 13:10:50.0061|Sealed block 13 (0x1ffaa0...dbf90b), diff: 340282366920938463463374607431768211454, tx count: 0 
el_1  | 2022-06-12 13:10:50.0094|Processed         13 |        3ms of   5,003ms, mgasps    0.00 total    3.80, tps    0.00 total   14.66, bps  390.53 total   87.95, recv queue 0, proc queue 0 
el_1  | 2022-06-12 13:10:50.2099|Error when handling ID 1, engine_forkchoiceUpdatedV1({
el_1  |   "headBlockHash": "0x4ebd4a511161d0ec70c6674395c8b9f06cfce7e120b12379c1ec2640f2fe7384",
el_1  |   "safeBlockHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
el_1  |   "finalizedBlockHash": "0x0000000000000000000000000000000000000000000000000000000000000000"
el_1  | },{
el_1  |   "timestamp": "0x62a5e5da",
el_1  |   "prevRandao": "0x67e28a0a07971c8e85a8e8c6e19f8ef08835aa3cb7409f671915f0b66e6e4ed6",
el_1  |   "suggestedFeeRecipient": "0x0000000000000000000000000000000000000000"
el_1  | }) | {"jsonrpc":"2.0","error":{"code":-32601,"message":"Method engine_forkchoiceUpdatedV1 is not supported"},"id":1} 
```

**To Reproduce**

- Clone https://github.com/dapplion/nethermind-tests/tree/cc1dcb21da9d9b68e991582c6ff9617df0f93e9a/nethermind-lodestar-m3-aura
- Run `docker-compose up`

**Expected behavior**

**Please please please: error immediately for any setup errors** :pray:  :pray:  :pray:

### PR DESCRIPTION
**Pull request by asdacap**
Resolves #4153

## Changes:
- Exit early on initialization step error
- Also make sure exit code is non-zero on error.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [X] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [X] No

Tested by setting no engine module at any url causing a validation error on Merge plugin startup. Application exited successfully.

### CODE REVIEW

File: src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs [hunk 1]
```diff
@@ -153,6 +153,7 @@ private void RunOneRoundOfInitialization(CancellationToken cancellationToken)
                     {
                         if (_logger.IsWarn) _logger.Warn(
                             $"Step {step.GetType().Name.PadRight(24)} failed after {stopwatch.ElapsedMilliseconds}ms");
+                        throw t.Exception;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs:156]: You know you are resetting call stack here?
Does this throw on some pool thread?

**Comment by LukaszRozmej** [comment]
This would fail on any exception, do we want that? I think not in genral. @MarekM25 @deffrian

**Comment by tanishqjasoria** [comment]
can we print out instructions on how the resolve the issue ( if its doable )
like when you are running two clients on the same machine, they both use port 30303, resulting in an error in starting the second client. In such cases we can tell the user what to do to resolve the issue. Something like this
`Network port 0.0.0.0:30303 is already in use, make sure that another instance of an Ethereum client is not running or change the port using the --port option`

**Comment by asdacap** [comment]
I think it should fail on any exception. This is on startup step BTW. Better that, then some part of the application just not working properly.

**Comment by asdacap** [comment]
> can we print out instructions on how the resolve the issue ( if its doable ) like when you are running two clients on the same machine, they both use port 30303, resulting in an error in starting the second client. In such cases we can tell the user what to do to resolve the issue. Something like this `Network port 0.0.0.0:30303 is already in use, make sure that another instance of an Ethereum client is not running or change the port using the --port option`

Good idea. I recommend creating another issue/PR for that. Should be fairly simple.

**Review by LukaszRozmej — COMMENTED** [review]
I'm not saying its wrong, but now we will fail on any plugin failing. So if we have a plugin that is like healthcheck or ethstats we close the whole app if it fails. Do we want that?

File: src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs:156]: Not entirely sure if will work but try `ExceptionDispatchInfo.Capture(t.Exception).Throw();`

**Review by LukaszRozmej — APPROVED** [review]
So it looks good to me. 
Two additional questions: Do we want to have some steps with MustInitialize to false and/or do we want to have some plugins that have something similar set to false?

Also before merging lets have a team discussion on daily about this again. Or maybe more people can revewiev it?

**Comment by asdacap** [comment]
The following step have mustinitialize set to false.
- InitializeNodeStats
- StartLogProducer
- StartMonitoring

**Pull request merged**